### ISSUE TITLE
USB MIDI support for other MIDI Controlers

### ISSUE THREAD

**Issue by tomassch** [MrBlueXav/Dekrispator#1]
Dear fellows,

I wanted to try out Dekrispator on my STM board, controlling it with my CME MIDI Controller.

I couldn't get any response from the controller. With GDB I could see that it would always give me "unrecognized device".

After some investigation, I found out that my controller has two interface descriptors. The first is some audio interface (maybe to send audio back from a hypothetical usb synth?) and the second is the real midi interface.

So I had to make changes in two places:
1)
.../Dekrispator/USB/usbh_conf.h around line 48
# define USBH_MAX_NUM_INTERFACES  1

Then obviously it would not decode my second interface. I just changed it to:
# define USBH_MAX_NUM_INTERFACES  2

2)
Now I had to tell Dekrispator to get data from that second interface:
.../Dekrispator/USB/MIDIstream_class/usbh_midi_core.c
In function USBH_MIDI_InterfaceInit, whenever there was a Itf_Desc[0][0], I changed to Itf_Desc[interfaceToUse][0]

interfaceToUse is a constant int = 1. 

Next step is making this selection automatic (don't want to mess with the support for the Korg Controller).

I am gonna request to pull after I reorganize my code (and maybe make the interface selection automatic). 

Very nice project to work with!

PS. Compiling all in Ubuntu 12.10, running make from the Release dir. Debugging with GDB/Emacs.

**Comment by tomassch** [comment]
Thanks a lot for porting the project to cube HAL.

I cloned the version 2, changed the paths in the makefile to relative paths, and it was running. Worked with my CME controller first time. If you allow me to push to v2, I can rename my dir to linuxRelease and upload it, so people can build from the command line as well.

One more thing: in MIDI_application.h I had to correct the case of a file in an include:

#include "usbh_midi.h" -> "usbh_MIDI.h"

What is using so much RAM? Synth related code?

Happy new year!

**Comment by MrBlueXav** [comment]
Thank you Tomas ! Happy new year too !
Include case issue now fixed.
Good news it works out of the box with your CME controller.

**Comment by MrBlueXav** [comment]
Let's go now on Dekrispator_v2 repository please !
About RAM usage, I've opened a new issue there...

**Comment by MrBlueXav** [comment]
@tomassch : OK for your LinuxRelease and thank you ! I've added you as a "collaborator".

**Comment by MrBlueXav** [comment]
Happy New Year to everybody here ! ;-))

**Comment by Reggi3** [comment]
Have you guys taken a look at the additions I've made to my fork?  I haven't cloned version 2 yet, as I would need to pull my code and re-add it....

**Comment by MrBlueXav** [comment]
Yes Reggie ! That's a nice work with good doc and many comments in the source files ! I can't test it at the moment however. I don't know much about endless encoders but maybe I could use the ones on my Prophet 08. Keep on the good work on your fork !

Thank you also for talking enthusiastically about Dekrispator on several famous forums !!!

My personal project is more of an autonomous machine making strange and sometimes impredictable sounds that you can tweak in real time....or not...., not really a fully controllable versatile synth.
I'd like to explore now polyphony with moving complex chords, evolving between harmonic/non-harmonic ratios. Still a kind of contemporary music box with controls.

**Comment by MrBlueXav** [comment]
I've also the "Juronstruktor" which is kind of sample player used to produce sequences of mixed up pieces of (french) dirty words... well that's a stupid thing I know but I love it. Should be on github soon.

http://youtu.be/spdb0K8gsQM?list=UUKbBi-lO1MtBKqolWJseSkQ

**Comment by Reggi3** [comment]
I really like the randomness of dekrispator but I found the randomness to be very extreme, I think I basically put in the ability to tame the randomness if you needed to.  I also started to explore the random sequencer, it's just a few functions away from being programmable on the fly.

Something else I was really interested in was making the software customisable, like being able to change the order of the effects units in the signal path, perhaps adding a 3rd oscillator/lfo combo, possibly even extra effects that haven't been done yet?

**Comment by MrBlueXav** [comment]
Better late than never... I've modified the USB host midi driver so it can automatically detect the midi interface in the connected device. I've used your different proposals, thank you guys !
Concerned files : usbh_midi_core.c/h and usbh_conf.h 
Could you, in case you have some spare time, test it with your various midi controllers ?

**Comment by mravenca** [comment]
Hello, does the Dekrispator now support multiple midi devices? I tried to connect synth controller(Akai LPD8) and keyboard(Keystation mini 32) via USB hub and they both do not turn on.
When I connect them directly to the USB on discovery board(first synth controller, then keyboard), they work.
